File: uniffi_bindgen/src/bindings/python/templates/TopLevelFunctionTemplate.py
{%- if func.is_async() %}

async def {{ func.name()|fn_name }}({%- call py::arg_list_decl(func) -%}):
    {%- call py::setup_args(func) %}
    # Get the `RustFuture`.
    rust_future = {% call py::to_ffi_call(func) %}
    future = None

    def trampoline() -> (FuturePoll, any):
        nonlocal rust_future

        {% match func.ffi_func().return_type() %}
        {% when Some with (return_type) %}
        polled_result = {{ return_type|ffi_type_name }}()
        is_ready = rust_call(_UniFFILib.{{ func.ffi_func().name() }}_poll, rust_future, future._future_ffi_waker(), ctypes.c_void_p(), ctypes.byref(polled_result))
        {% when None %}
        is_ready = rust_call(_UniFFILib.{{ func.ffi_func().name() }}_poll, rust_future, future._future_ffi_waker(), ctypes.c_void_p(), ctypes.c_void_p())
        {% endmatch %}

        if is_ready is True:
            result = {% match func.return_type() %}{% when Some with (return_type) %}{{ return_type|lift_fn }}(polled_result){% when None %}None{% endmatch %}

            return (FuturePoll.DONE, result)
        else:
            return (FuturePoll.PENDING, None)

    # Create our own Python `Future` and poll it.
    future = Future(trampoline)
    future.init()

    # Let's wait on it.
    result = await future

    # Drop the `rust_future`.
    rust_call(_UniFFILib.{{ func.ffi_func().name() }}_drop, rust_future)

    return result
{%- else %}
{%- match func.return_type() -%}
{%- when Some with (return_type) %}

def {{ func.name()|fn_name }}({%- call py::arg_list_decl(func) -%}):
    {%- call py::setup_args(func) %}
    return {{ return_type|lift_fn }}({% call py::to_ffi_call(func) %})
{% when None %}

def {{ func.name()|fn_name }}({%- call py::arg_list_decl(func) -%}):
    {%- call py::setup_args(func) %}
    {% call py::to_ffi_call(func) %}
{% endmatch %}
{%- endif %}

File: uniffi_bindgen/src/bindings/kotlin/templates/TopLevelFunctionTemplate.kt
{%- if func.is_async() %}
{%- match func.throws_type() -%}
{%- when Some with (throwable) %}
@Throws({{ throwable|type_name }}::class)
{%- else -%}
{%- endmatch %}
suspend fun {{ func.name()|fn_name }}({%- call kt::arg_list_decl(func) -%}){% match func.return_type() %}{% when Some with (return_type) %}: {{ return_type|type_name }}{% when None %}{% endmatch %} {
    class Waker: RustFutureWaker {
        override fun callback(envCStructure: RustFutureWakerEnvironmentCStructure?) {
            if (envCStructure == null) {
                return;
            }

            val hash = envCStructure.hash
            val env = _UniFFILib.FUTURE_WAKER_ENVIRONMENTS.remove(hash)

            if (env == null) {
                return;
            }

            env.coroutineScope.launch {
                @Suppress("UNCHECKED_CAST")
                val continuation = {% match func.return_type() -%}
                {%- when Some with (return_type) -%}
                    env.continuation as Continuation<{{ return_type|type_name }}>
                {%- when None -%}
                    env.continuation as Continuation<Unit>
                {%- endmatch %}
                val polledResult = {% match func.ffi_func().return_type() -%}
                {%- when Some with (return_type) -%}
                    {{ return_type|type_ffi_lowered }}
                {%- when None -%}
                    Pointer
                {%- endmatch %}ByReference()

                try {
                    val isReady = {% match func.throws_type() -%}
                    {%- when Some with (error) -%}
                        rustCallWithError({{ error|type_name }})
                    {%- when None -%}
                        rustCall()
                    {%- endmatch %}
                    { _status ->
                        _UniFFILib.INSTANCE.{{ func.ffi_func().name() }}_poll(
                            env.rustFuture,
                            env.waker,
                            env.selfAsCStructure,
                            polledResult,
                            _status
                        )
                    }

                    if (isReady) {
                        continuation.resume(
                        {% match func.return_type() -%}
                        {%- when Some with (return_type) -%}
                            {{ return_type|lift_fn}}(polledResult.getValue())
                        {%- when None -%}
                            Unit
                        {%- endmatch %}
                        )

                        rustCall() { _status ->
                            _UniFFILib.INSTANCE.{{ func.ffi_func().name() }}_drop(env.rustFuture, _status)
                        }
                    } else {
                        _UniFFILib.FUTURE_WAKER_ENVIRONMENTS.put(hash, env)
                    }
                } catch (exception: Exception) {
                    continuation.resumeWithException(exception)

                    rustCall() { _status ->
                        _UniFFILib.INSTANCE.{{ func.ffi_func().name() }}_drop(env.rustFuture, _status)
                    }
                }
            }
        }
    }

    val result: {% match func.return_type() %}{% when Some with (return_type) %}{{ return_type|type_name }}{% when None %}Unit{% endmatch %}

    coroutineScope {
        result = suspendCoroutine<{% match func.return_type() %}{% when Some with (return_type) %}{{ return_type|type_name }}{% when None %}Unit{% endmatch %}> { continuation ->
            val rustFuture = {% call kt::to_ffi_call(func) %}

            val env = RustFutureWakerEnvironment(rustFuture, continuation, Waker(), RustFutureWakerEnvironmentCStructure(), this)
            val envHash = env.hashCode()
            env.selfAsCStructure.hash = envHash

            _UniFFILib.FUTURE_WAKER_ENVIRONMENTS.put(envHash, env)

            val waker = Waker()
            waker.callback(env.selfAsCStructure)
        }
    }

    return result
}

{%- else %}
{%- match func.throws_type() -%}
{%- when Some with (throwable) %}
@Throws({{ throwable|type_name }}::class)
{%- else -%}
{%- endmatch -%}
{%- match func.return_type() -%}
{%- when Some with (return_type) %}
fun {{ func.name()|fn_name }}({%- call kt::arg_list_decl(func) -%}): {{ return_type|type_name }} {
    return {{ return_type|lift_fn }}({% call kt::to_ffi_call(func) %})
}

{% when None %}

fun {{ func.name()|fn_name }}({% call kt::arg_list_decl(func) %}) =
    {% call kt::to_ffi_call(func) %}
{% endmatch %}
{%- endif %}
